File: packages/maencof/skills/craft-skill/scripts/package_skill.py
#!/usr/bin/env python3
"""
Enhanced Skill Packager - Creates distributable skill packages

Usage:
    package_skill.py <skill-path> [output-dir]

Features:
- Integrated enhanced_validator.py validation
- Version metadata inclusion
- CHANGELOG.md auto-update support
- Detailed packaging logs
"""

import argparse
import hashlib
import json
import subprocess
import sys
import zipfile
from datetime import datetime
from pathlib import Path
from typing import Dict, List


def calculate_checksum(file_path: Path) -> str:
    """Calculate SHA256 checksum of a file"""
    sha256 = hashlib.sha256()
    with open(file_path, 'rb') as f:
        for chunk in iter(lambda: f.read(4096), b''):
            sha256.update(chunk)
    return sha256.hexdigest()


def extract_metadata(skill_path: Path) -> Dict:
    """Extract metadata from SKILL.md frontmatter"""
    skill_md = skill_path / "SKILL.md"

    if not skill_md.exists():
        raise ValueError("SKILL.md not found")

    import re
    content = skill_md.read_text()

    match = re.match(r'^---\n(.*?)\n---', content, re.DOTALL)
    if not match:
        raise ValueError("Invalid YAML frontmatter")

    frontmatter = match.group(1)

    metadata = {}
    for line in frontmatter.split('\n'):
        if ':' in line:
            key, value = line.split(':', 1)
            metadata[key.strip()] = value.strip()

    return metadata


def validate_skill(skill_path: Path) -> bool:
    """Run enhanced_validator.py on the skill"""
    script_dir = Path(__file__).parent
    validator = script_dir / "enhanced_validator.py"

    if not validator.exists():
        print("⚠️  Warning: enhanced_validator.py not found, skipping validation")
        return True

    print("🔍 Running validation...")
    result = subprocess.run(
        [sys.executable, str(validator), str(skill_path)],
        capture_output=True,
        text=True
    )

    print(result.stdout)

    return result.returncode == 0


def get_skill_structure_info(skill_path: Path) -> Dict:
    """Analyze skill structure"""
    info = {
        "has_reference_md": (skill_path / "reference.md").exists(),
        "has_examples_md": (skill_path / "examples.md").exists(),
        "has_knowledge": (skill_path / "knowledge").exists(),
        "has_docs": (skill_path / "docs").exists(),
        "scripts_count": len(list((skill_path / "scripts").glob("*.py"))) if (skill_path / "scripts").exists() else 0
    }
    return info


def create_install_instructions(metadata: Dict, structure: Dict, skill_name: str) -> str:
    """Generate INSTALL.md content"""
    version = metadata.get('version', '1.0.0')

    content = f"""# Installation Instructions for {skill_name} v{version}

## Quick Install

1. Extract the zip file:
   ```bash
   unzip {skill_name}.zip
   ```

2. Move to Claude Code skills directory:
   ```bash
   mv {skill_name} ~/.claude/skills/
   ```

3. Verify installation:
   ```bash
   python ~/.claude/skills/{skill_name}/scripts/enhanced_validator.py ~/.claude/skills/{skill_name}
   ```

## Structure

This skill includes:
"""

    if structure['has_reference_md']:
        content += "- ✅ reference.md - Detailed documentation\n"
    if structure['has_examples_md']:
        content += "- ✅ examples.md - Usage examples\n"
    if structure['has_knowledge']:
        content += "- ✅ knowledge/ - Deep-dive topics\n"
    if structure['has_docs']:
        content += "- ✅ docs/ - Additional guides\n"

    content += f"- ✅ {structure['scripts_count']} automation script(s)\n"

    content += """
## Dependencies

Check the skill's requirements.txt (if present) for dependencies:
```bash
cd ~/.claude/skills/{skill_name}
pip install -r requirements.txt  # if requirements.txt exists
```

## Verification

After installation, ask Claude:
> "Use the {skill_name} skill to [example task]"

Claude should recognize and use the skill automatically.

## Troubleshooting

If Claude doesn't recognize the skill:
1. Check file permissions: `ls -la ~/.claude/skills/{skill_name}`
2. Verify SKILL.md exists and has valid YAML frontmatter
3. Restart Claude Code application

## Uninstallation

To remove this skill:
```bash
rm -rf ~/.claude/skills/{skill_name}
```
"""

    return content.format(skill_name=skill_name)


def package_skill(skill_path: Path, output_dir: Path) -> bool:
    """Create skill package"""
    skill_name = skill_path.name

    print(f"📦 Packaging skill: {skill_name}")

    # Validate first
    if not validate_skill(skill_path):
        print("❌ Validation failed. Fix issues before packaging.")
        return False

    # Extract metadata
    try:
        metadata = extract_metadata(skill_path)
        print(f"📋 Version: {metadata.get('version', 'unknown')}")
    except Exception as e:
        print(f"❌ Error extracting metadata: {e}")
        return False

    # Analyze structure
    structure = get_skill_structure_info(skill_path)

    # Create output directory
    output_dir.mkdir(parents=True, exist_ok=True)

    # Create package metadata
    package_metadata = {
        "name": skill_name,
        "version": metadata.get('version', '1.0.0'),
        "complexity": metadata.get('complexity', 'unknown'),
        "created": metadata.get('created', ''),
        "packaged": datetime.now().isoformat(),
        "requires": {
            "claude_code_version": ">=1.0.0",
            "python_version": ">=3.8"
        },
        "structure": structure
    }

    # Create zip file
    zip_path = output_dir / f"{skill_name}.zip"

    print(f"🗜️  Creating package: {zip_path}")

    with zipfile.ZipFile(zip_path, 'w', zipfile.ZIP_DEFLATED) as zipf:
        # Add all skill files
        for file in skill_path.rglob('*'):
            if file.is_file():
                arcname = f"{skill_name}/{file.relative_to(skill_path)}"
                zipf.write(file, arcname)
                print(f"  + {file.relative_to(skill_path)}")

        # Add metadata
        metadata_json = json.dumps(package_metadata, indent=2)
        zipf.writestr(f"{skill_name}/.skill-metadata.json", metadata_json)
        print(f"  + .skill-metadata.json")

        # Add INSTALL.md
        install_content = create_install_instructions(metadata, structure, skill_name)
        zipf.writestr("INSTALL.md", install_content)
        print(f"  + INSTALL.md")

    # Calculate checksum
    checksum = calculate_checksum(zip_path)
    package_metadata['checksum'] = f"sha256:{checksum}"

    # Save metadata separately
    metadata_path = output_dir / f"{skill_name}.metadata.json"
    with open(metadata_path, 'w') as f:
        json.dump(package_metadata, f, indent=2)

    print(f"\n✅ Package created successfully!")
    print(f"📦 Package: {zip_path}")
    print(f"📋 Metadata: {metadata_path}")
    print(f"🔐 Checksum: {checksum[:16]}...")

    return True


def main():
    parser = argparse.ArgumentParser(description="Package a skill for distribution")
    parser.add_argument("skill_path", help="Path to skill directory")
    parser.add_argument("output_dir", nargs='?', default="./dist",
                       help="Output directory (default: ./dist)")

    args = parser.parse_args()

    skill_path = Path(args.skill_path).resolve()
    output_dir = Path(args.output_dir).resolve()

    if not skill_path.exists():
        print(f"❌ Error: Skill path not found: {skill_path}")
        sys.exit(1)

    success = package_skill(skill_path, output_dir)
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: packages/maencof/skills/craft-skill/scripts/package_skill.mjs
#!/usr/bin/env node
/**
 * Enhanced Skill Packager - Creates distributable skill packages
 *
 * Usage:
 *     node package_skill.mjs <skill-path> [output-dir]
 *
 * Features:
 * - Integrated enhanced_validator.mjs validation
 * - Version metadata inclusion
 * - CHANGELOG.md auto-update support
 * - Detailed packaging logs
 */

import { parseArgs } from 'node:util';
import { execFileSync, execSync } from 'node:child_process';
import {
  existsSync, readFileSync, readdirSync, mkdirSync, writeFileSync, rmSync, mkdtempSync, cpSync,
} from 'node:fs';
import { createHash } from 'node:crypto';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { tmpdir } from 'node:os';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function calculateChecksum(filePath) {
  const hash = createHash('sha256');
  const content = readFileSync(filePath);
  hash.update(content);
  return hash.digest('hex');
}

function extractMetadata(skillPath) {
  const skillMd = path.join(skillPath, 'SKILL.md');
  if (!existsSync(skillMd)) {
    throw new Error('SKILL.md not found');
  }

  const content = readFileSync(skillMd, 'utf-8');
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) {
    throw new Error('Invalid YAML frontmatter');
  }

  const frontmatter = match[1];
  const metadata = {};

  for (const line of frontmatter.split('\n')) {
    const colonIdx = line.indexOf(':');
    if (colonIdx !== -1) {
      const key = line.slice(0, colonIdx).trim();
      const value = line.slice(colonIdx + 1).trim();
      metadata[key] = value;
    }
  }

  return metadata;
}

function validateSkill(skillPath) {
  const validator = path.join(__dirname, 'enhanced_validator.mjs');

  if (!existsSync(validator)) {
    console.log('⚠️  Warning: enhanced_validator.mjs not found, skipping validation');
    return true;
  }

  console.log('🔍 Running validation...');
  try {
    const output = execFileSync('node', [validator, skillPath], { encoding: 'utf-8' });
    console.log(output);
    return true;
  } catch (e) {
    if (e.stdout) console.log(e.stdout);
    return false;
  }
}

function getSkillStructureInfo(skillPath) {
  const scriptsDir = path.join(skillPath, 'scripts');
  let scriptsCount = 0;
  if (existsSync(scriptsDir)) {
    scriptsCount = readdirSync(scriptsDir).filter(
      (f) => f.endsWith('.py') || f.endsWith('.mjs') || f.endsWith('.js')
    ).length;
  }

  return {
    has_reference_md: existsSync(path.join(skillPath, 'reference.md')),
    has_examples_md: existsSync(path.join(skillPath, 'examples.md')),
    has_knowledge: existsSync(path.join(skillPath, 'knowledge')),
    has_docs: existsSync(path.join(skillPath, 'docs')),
    scripts_count: scriptsCount,
  };
}

function createInstallInstructions(metadata, structure, skillName) {
  const versionSuffix = metadata.version ? ` v${metadata.version}` : '';

  let content = `# Installation Instructions for ${skillName}${versionSuffix}

## Quick Install

1. Extract the zip file:
   \`\`\`bash
   unzip ${skillName}.zip
   \`\`\`

2. Move to Claude Code skills directory:
   \`\`\`bash
   mv ${skillName} ~/.claude/skills/
   \`\`\`

3. Verify installation:
   \`\`\`bash
   node ~/.claude/skills/${skillName}/scripts/enhanced_validator.mjs ~/.claude/skills/${skillName}
   \`\`\`

## Structure

This skill includes:
`;

  if (structure.has_reference_md) content += '- ✅ reference.md - Detailed documentation\n';
  if (structure.has_examples_md) content += '- ✅ examples.md - Usage examples\n';
  if (structure.has_knowledge) content += '- ✅ knowledge/ - Deep-dive topics\n';
  if (structure.has_docs) content += '- ✅ docs/ - Additional guides\n';
  content += `- ✅ ${structure.scripts_count} automation script(s)\n`;

  content += `
## Verification

After installation, ask Claude:
> "Use the ${skillName} skill to [example task]"

Claude should recognize and use the skill automatically.

## Troubleshooting

If Claude doesn't recognize the skill:
1. Check file permissions: \`ls -la ~/.claude/skills/${skillName}\`
2. Verify SKILL.md exists and has valid YAML frontmatter
3. Restart Claude Code application

## Uninstallation

To remove this skill:
\`\`\`bash
rm -rf ~/.claude/skills/${skillName}
\`\`\`
`;

  return content;
}

function collectFiles(dir, base) {
  const results = [];
  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    const relPath = path.relative(base, fullPath);
    if (entry.isFile()) {
      results.push({ fullPath, relPath });
    } else if (entry.isDirectory()) {
      results.push(...collectFiles(fullPath, base));
    }
  }
  return results;
}

function packageSkill(skillPath, outputDir) {
  const skillName = path.basename(skillPath);

  console.log(`📦 Packaging skill: ${skillName}`);

  if (!validateSkill(skillPath)) {
    console.log('❌ Validation failed. Fix issues before packaging.');
    return false;
  }

  let metadata;
  try {
    metadata = extractMetadata(skillPath);
    if (metadata.version) {
      console.log(`📋 Version: ${metadata.version}`);
    }
  } catch (e) {
    console.error(`❌ Error extracting metadata: ${e.message}`);
    return false;
  }

  const structure = getSkillStructureInfo(skillPath);

  mkdirSync(outputDir, { recursive: true });

  const packageMetadata = {
    name: skillName,
    packaged: new Date().toISOString(),
    requires: {
      claude_code_version: '>=1.0.0',
      node_version: '>=18.0',
    },
    structure,
  };

  for (const field of ['version', 'complexity', 'created']) {
    if (metadata[field]) {
      packageMetadata[field] = metadata[field];
    }
  }

  // Stage files in a temp directory for zip creation
  const tmpDir = mkdtempSync(path.join(tmpdir(), 'skill-pkg-'));
  const stageDir = path.join(tmpDir, skillName);

  try {
    // Copy skill files to staging
    cpSync(skillPath, stageDir, { recursive: true });

    // Add metadata
    writeFileSync(
      path.join(stageDir, '.skill-metadata.json'),
      JSON.stringify(packageMetadata, null, 2)
    );
    console.log('  + .skill-metadata.json');

    // Add INSTALL.md at top level of staging
    const installContent = createInstallInstructions(metadata, structure, skillName);
    writeFileSync(path.join(tmpDir, 'INSTALL.md'), installContent);
    console.log('  + INSTALL.md');

    // Log files being packaged
    const files = collectFiles(stageDir, stageDir);
    for (const f of files) {
      console.log(`  + ${f.relPath}`);
    }

    // Create zip
    const zipPath = path.join(outputDir, `${skillName}.zip`);
    console.log(`\n🗜️  Creating package: ${zipPath}`);

    execSync(`cd "${tmpDir}" && zip -r "${zipPath}" "${skillName}" INSTALL.md`, {
      stdio: 'pipe',
    });

    // Calculate checksum
    const checksum = calculateChecksum(zipPath);
    packageMetadata.checksum = `sha256:${checksum}`;

    // Save metadata separately
    const metadataPath = path.join(outputDir, `${skillName}.metadata.json`);
    writeFileSync(metadataPath, JSON.stringify(packageMetadata, null, 2));

    console.log('\n✅ Package created successfully!');
    console.log(`📦 Package: ${zipPath}`);
    console.log(`📋 Metadata: ${metadataPath}`);
    console.log(`🔐 Checksum: ${checksum.slice(0, 16)}...`);

    return true;
  } finally {
    rmSync(tmpDir, { recursive: true, force: true });
  }
}

// --- Main ---

const { positionals } = parseArgs({
  allowPositionals: true,
  strict: true,
});

if (positionals.length === 0) {
  console.error('Usage: node package_skill.mjs <skill-path> [output-dir]');
  process.exit(1);
}

const skillPath = path.resolve(positionals[0]);
const outputDir = path.resolve(positionals[1] || './dist');

if (!existsSync(skillPath)) {
  console.error(`❌ Error: Skill path not found: ${skillPath}`);
  process.exit(1);
}

const success = packageSkill(skillPath, outputDir);
process.exit(success ? 0 : 1);
